src/ucp/core/ucp_ep.c

```diff
@@ -351,14 +352,20 @@ ucp_ep_create_api_to_worker_addr(ucp_worker_h worker,
         goto out;
     }
 
-    /* check if there is already an unconnected internal endpoint to the same
-     * destination address
+    /* Check if there is already an unconnected internal endpoint to the same
+     * destination address.
+     * In case of loopback connection, search the hash table for an endpoint with
+     * even/odd matching, so that every 2 endpoints connected to the local worker
+     * with be paired to each other.
+     * Note that if a loopback endpoint had UCP_EP_PARAMS_FLAGS_NO_LOOPBACK flag,
+     * it will not be added to ep_match as unexpected. This is because dest_ep_ptr
+     * will be initialized, so a WIREUP_REQUEST (if sent) will have dst_ep != 0,
+     * so ucp_wireup_process_request() will not create unexpected ep in ep_match.
      */
     conn_sn = ucp_ep_match_get_next_sn(&worker->ep_match_ctx, remote_address.uuid);
     ep = ucp_ep_match_retrieve_unexp(&worker->ep_match_ctx, remote_address.uuid,
-                                     conn_sn);
+                                     conn_sn ^ (remote_address.uuid == worker->uuid));
```

Comment:
minor: really unusual construction: bitwise operation between number and 1-bit (comparing result) integer :)

Reply:
yeah, this is the reason i added comment above :)